### PR TITLE
fix invalid read detected by valgrind

### LINKED ISSUE #3525: [R-package] invalid read detected by valgrind
Based on the result of a recent run of the valgrind job (https://github.com/microsoft/LightGBM/runs/1349685476?check_suite_focus=true), it looks like recent commits have introduced an illegal read.

```text
==2055== Invalid read of size 1
==2055==    at 0x15335D0F: LightGBM::LeafSplits::Init(float const*, float const*) (leaf_splits.hpp:74)
==2055==    by 0x15329D9E: LightGBM::SerialTreeLearner::BeforeTrain() (serial_tree_learner.cpp:264)
==2055==    by 0x153296B3: LightGBM::SerialTreeLearner::Train(float const*, float const*) (serial_tree_learner.cpp:166)
==2055==    by 0x151A0A56: LightGBM::GBDT::TrainOneIter(float const*, float const*) (gbdt.cpp:379)
==2055==    by 0x15458612: LightGBM::Booster::TrainOneIter(float const*, float const*) (c_api.cpp:364)
==2055==    by 0x15449C4E: LGBM_BoosterUpdateOneIterCustom (c_api.cpp:1634)
==2055==    by 0x1546C469: LGBM_BoosterUpdateOneIterCustom_R (lightgbm_R.cpp:417)
==2055==    by 0x4941BDA: R_doDotCall (dotcode.c:610)
==2055==    by 0x494CCC6: do_dotcall (dotcode.c:1281)
==2055==    by 0x499FB01: bcEval (eval.c:7078)
==2055==    by 0x498B67F: Rf_eval (eval.c:727)
==2055==    by 0x498E414: R_execClosure (eval.c:1895)
==2055==  Address 0x8570974 is 260 bytes inside a block of size 1,512 free'd
==2055==    at 0x483CFBF: operator delete(void*) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2055==    by 0x151AA06F: std::default_delete<LightGBM::Config>::operator()(LightGBM::Config*) const (unique_ptr.h:85)
==2055==    by 0x151AA6FA: std::__uniq_ptr_impl<LightGBM::Config, std::default_delete<LightGBM::Config> >::reset(LightGBM::Config*) (unique_ptr.h:182)
==2055==    by 0x151A93C6: std::unique_ptr<LightGBM::Config, std::default_delete<LightGBM::Config> >::reset(LightGBM::Config*) (unique_ptr.h:456)
==2055==    by 0x151A3644: LightGBM::GBDT::ResetConfig(LightGBM::Config const*) (gbdt.cpp:753)
==2055==    by 0x15457E9C: LightGBM::Booster::ResetConfig(char const*) (c_api.cpp:329)
==2055==    by 0x15449721: LGBM_BoosterResetParameter (c_api.cpp:1592)
==2055==    by 0x1546BF4F: LGBM_BoosterResetParameter_R (lightgbm_R.cpp:381)
==2055==    by 0x4941B4C: R_doDotCall (dotcode.c:604)
==2055==    by 0x494CCC6: do_dotcall (dotcode.c:1281)
==2055==    by 0x499FB01: bcEval (eval.c:7078)
==2055==    by 0x498B67F: Rf_eval (eval.c:727)
```

I'm marking this `blocking` for #3484 , because I think CRAN will reject the package if we don't fix this.

Part of the issue here is that the check in https://github.com/microsoft/LightGBM/blob/cf69591b68c68a78ea4064e6941b8270b8a00eef/.ci/test_r_package_valgrind.sh for whether the valgrind test passed / failed is not strict enough. It currently only checks for the number of bytes lost, but should also be looking for "invalid read" and "invalid write" as well. I can fix that part.

### PR DESCRIPTION
**Pull request by guolinke**
to fix #3525

### CODE REVIEW

File: src/treelearner/leaf_splits.hpp [hunk 1]
```diff
@@ -22,8 +22,13 @@ namespace LightGBM {
 class LeafSplits {
  public:
   LeafSplits(data_size_t num_data, const Config* config)
-    :config_(config), num_data_in_leaf_(num_data), num_data_(num_data),
+      : deterministic_(false),
```
Comment (guolinke) [inline src/treelearner/leaf_splits.hpp:25]: /gha run r-valgrind

Comment (jameslamb) [reply src/treelearner/leaf_splits.hpp:25]: looks like your changes fixed it!

https://github.com/microsoft/LightGBM/runs/1350825291?check_suite_focus=true

The only issues left there are the ones we already knew about as of #3443 , which we think are false positives.

**Comment by jameslamb** [comment]
Looks ok to me, but I don't think I'm qualified to give this a :heavy_check_mark:

**Pull request merged**